webapp/utils/channel_utils.jsx

```diff
@@ -0,0 +1,150 @@
+        favoriteChannels,
+        publicChannels: notFavoriteChannels.filter(isOpenChannel),
+        privateChannels: notFavoriteChannels.filter(isPrivateChannel),
+        directChannels: directChannels.filter(isConnectedToTeamMember),
+        directNonTeamChannels: directChannels.filter(not(isConnectedToTeamMember))
+    };
+}
+
+export function isFavoriteChannel(channel) {
+    return PreferenceStore.getBool(Preferences.CATEGORY_FAVORITE_CHANNEL, channel.id);
+}
+
+export function isOpenChannel(channel) {
+    return channel.type === Constants.OPEN_CHANNEL;
+}
+
+export function isPrivateChannel(channel) {
+    return channel.type === Constants.PRIVATE_CHANNEL;
+}
+
+export function isDirectChannel(channel) {
+    return channel.type === Constants.DM_CHANNEL;
+}
+
+export function isDirectChannelVisible(channel) {
+    const channelId = Utils.getUserIdFromChannelName(channel);
+
+    return PreferenceStore.getBool(Preferences.CATEGORY_DIRECT_CHANNEL_SHOW, channelId);
+}
+
+export function completeDirectChannelInfo(channel) {
+    if (!isDirectChannel(channel)) {
+        return channel;
+    }
+
+    const dmChannelClone = JSON.parse(JSON.stringify(channel));
+    const teammateId = Utils.getUserIdFromChannelName(channel);
+
+    return Object.assign(dmChannelClone, {
+        display_name: Utils.displayUsername(teammateId),
+        teammate_id: teammateId,
+        status: UserStore.getStatus(teammateId) || 'offline'
+    });
+}
+
+export function sortChannelsByDisplayName(a, b) {
+    const locale = LocalizationStore.getLocale();
+
+    return buildDisplayNameAndTypeComparable(a).localeCompare(buildDisplayNameAndTypeComparable(b), locale, {numeric: true});
+}
+
+/*
+ * not exported helpers
+ */
+
+function createMissingDirectChannels(channels) {
+    const directChannelsDisplayPreferences = PreferenceStore.getCategory(Preferences.CATEGORY_DIRECT_CHANNEL_SHOW);
+
+    return Array.
+        from(directChannelsDisplayPreferences).
+        filter((entry) => entry[1] === 'true').
+        map((entry) => entry[0]).
+        filter((teammateId) => !channels.some(Utils.isDirectChannelForUser.bind(null, teammateId))).
+        map(createFakeChannelCurried(UserStore.getCurrentId()));
+}
+
+function createFakeChannel(userId, otherUserId) {
+    return {
+        name: Utils.getDirectChannelName(userId, otherUserId),
+        last_post_at: 0,
+        total_msg_count: 0,
+        type: Constants.DM_CHANNEL,
+        fake: true
+    };
+}
+
+function createFakeChannelCurried(userId) {
+    return (otherUserId) => createFakeChannel(userId, otherUserId);
+}
+
+function isTeamMember(userId) {
+    return TeamStore.hasActiveMemberInTeam(TeamStore.getCurrentId(), userId);
+}
+
+function extractTeammateId(channel) {
+    return channel.teammate_id;
+}
+
+function not(f) {
+    return (...args) => {
+        return !f(...args);
+    };
+}
+
+function andX(...fns) {
+    return (...args) => fns.every((f) => f(...args));
+}
+
+function compose(...fns) {
```

Comment:
The core team talked about your usage of functional constructs here and are concerned about maintainability. 
While I like functional languages, JavaScript is unfortunately not one. Hacking functional constructs into the language like this is confusing for many developers. Since Mattermost is a community effort we want to keep it as welcoming as we can.

Therefore can we ask you to replace these functional constructs with more idiomatic JavaScript?
For example. Your usage of this compose() fucntion could be replace simply by:
```javascript
const isConnectedToTeamMember = (arg) => {
    isTeamMember(extractTeamId(arg))
}
```